File: eam/bridge/persistence/history.py
"""Read native CLI history without copying it to EAM or issuing AI requests.

Bounded JSON lines, SQLite rows and display pages. Schema mismatches fail visibly.
"""
from contextlib import contextmanager
import json
from pathlib import Path
import sqlite3
import sys

ROW_LIMIT = 8 * 1024 * 1024
LIST_LIMIT = 300


@contextmanager
def database(path):
    con = sqlite3.connect(Path(path).resolve().as_uri() + '?mode=ro', uri=True, timeout=2)
    try:
        con.execute('PRAGMA query_only=ON')
        yield con
    finally:
        con.close()


def json_lines(path):
    with open(path, 'rb') as source:
        while True:
            line = source.readline(ROW_LIMIT + 1)
            if not line:
                break
            if len(line) > ROW_LIMIT:
                raise ValueError('Native history item exceeds 8 MiB; refusing an incomplete view')
            try:
                value = json.loads(line)
            except ValueError:
                # A writer may not have finished the last JSON line yet.
                if not line.endswith(b'\n'):
                    break
                raise ValueError('Unrecognized native JSONL record')
            if isinstance(value, dict):
                yield value


def text_parts(content):
    if isinstance(content, str):
        return content
    if not isinstance(content, list):
        return ''
    return '\n'.join(item['text'] for item in content
                     if isinstance(item, dict) and item.get('type') in ('text', 'input_text', 'output_text')
                     and isinstance(item.get('text'), str))


def messages(entry):
    if entry['provider'] == 'Codex' and entry.get('mode') == 'paginated':
        with database(entry['history_db']) as con:
            rows = con.execute('SELECT length(item_json), substr(item_json,1,?), item_type FROM thread_items '
                               'WHERE thread_id=? AND item_type IN (?,?) ORDER BY rollout_ordinal',
                               (ROW_LIMIT + 1, entry['id'], 'userMessage', 'agentMessage'))
            for length, raw, kind in rows:
                if length > ROW_LIMIT or len(raw.encode('utf-8')) > ROW_LIMIT:
                    raise ValueError('Native history item exceeds 8 MiB')
                item = json.loads(raw)
                text = item.get('text', '') if kind == 'agentMessage' else text_parts(item.get('content'))
                if text:
                    yield ('assistant' if kind == 'agentMessage' else 'user'), text
        return
    for item in json_lines(entry['path']):
        if entry['provider'] == 'Claude':
            if item.get('sessionId') not in (None, entry['id']):
                continue
            if item.get('isSidechain') or item.get('isMeta') or item.get('type') not in ('user', 'assistant'):
                continue
            message = item.get('message', {})
        else:
            if item.get('type') != 'response_item':
                continue
            message = item.get('payload', {})
            if message.get('type') != 'message':
                continue
        if not isinstance(message, dict) or message.get('role') not in ('user', 'assistant'):
            continue
        text = text_parts(message.get('content'))
        if text:
            yield message['role'], text


def catalog(provider, root, directory=None):
    root = Path(root).expanduser()
    entries = []
    if provider == 'Codex':
        # Require an identified schema version rather than silently reading an
        # unrelated DB after a CLI migration.
        with database(root / 'state_5.sqlite') as con:
            rows = con.execute('SELECT id,cwd,title,rollout_path,history_mode,updated_at FROM threads '
                               'WHERE (? IS NULL OR rtrim(cwd,\'/\')=rtrim(?,\'/\')) '
                               'ORDER BY updated_at DESC LIMIT ?', (directory, directory, LIST_LIMIT + 1))
            for identity, cwd, title, path, mode, updated in rows:
                if mode not in ('legacy', 'paginated'):
                    raise ValueError('Unsupported Codex history mode: ' + str(mode))
                entries.append(dict(provider=provider, id=identity, directory=cwd,
                                    title=(title or '')[:100], path=path, mode=mode, updated=updated,
                                    history_db=str(root / 'thread_history_1.sqlite')))
    elif provider == 'Claude':
        projects = root / 'projects'
        if not projects.is_dir():
            raise ValueError('Claude projects directory not found')
        # Iterate instead of materializing every path. Bound discovery explicitly.
        count = 0
        for project in projects.iterdir():
            if not project.is_dir():
                continue
            for path in project.glob('*.jsonl'):
                count += 1
                if count > 10000:
                    raise ValueError('More than 10000 Claude histories; narrow the configured root')
                cwd = title = None
                for index, item in enumerate(json_lines(path)):
                    cwd = cwd or item.get('cwd')
                    if item.get('type') == 'user' and not item.get('isMeta'):
                        title = text_parts(item.get('message', {}).get('content'))[:100]
                    if (cwd and title) or index >= 100:
                        break
                if directory and (not cwd or Path(cwd).resolve() != Path(directory).resolve()):
                    continue
                entries.append(dict(provider=provider, id=path.stem, directory=cwd or '', title=title or path.stem,
                                    path=str(path), updated=path.stat().st_mtime))
                entries.sort(key=lambda e: e['updated'], reverse=True)
                del entries[LIST_LIMIT+1:]
    else:
        raise ValueError('Unsupported provider')
    return dict(entries=entries[:LIST_LIMIT], truncated=len(entries) > LIST_LIMIT)


def page(entry, offset, limit, prompt=False):
    limit = max(128, min(int(limit), 65536))
    offset = None if offset is None else max(0, int(offset))
    total = 0
    shown = ''
    for role, text in messages(entry):
        if prompt:
            if role == 'user':
                start = 0 if offset is None else min(offset, len(text))
                shown = text[start:start+limit]
                total = len(text)
            continue
        block = ('\n## 사용자\n\n' if role == 'user' else '\n## 응답\n\n') + text + '\n'
        if offset is None:
            shown = (shown + block[-limit:])[-limit:]
        else:
            a = max(0, offset - total)
            b = min(len(block), offset + limit - total)
            if b > a:
                shown += block[a:b]
        total += len(block)
    start = (0 if offset is None else min(offset, total)) if prompt else max(0, total-len(shown)) if offset is None else min(offset, total)
    return dict(text=shown, start=start, end=start+len(shown), total=total,
                provider=entry['provider'], id=entry['id'], prompt=prompt)


if __name__ == '__main__':
    try:
        request = json.loads(sys.stdin.buffer.read(65537))
        if request['action'] == 'list':
            result = catalog(request['provider'], request['root'], request.get('directory'))
        else:
            result = page(request['entry'], request.get('offset'), request.get('limit',65536), request.get('prompt',False))
        print(json.dumps(result, ensure_ascii=False))
    except (OSError, ValueError, KeyError, TypeError, sqlite3.Error) as exc:
        print(json.dumps(dict(error=str(exc)), ensure_ascii=False))
        sys.exit(1)

File: eam/native/src/history.rs
use crate::common::*;
use rusqlite::{params, Connection, OpenFlags};
use serde_json::{json, Value};
use std::{
    fs::{self, File},
    io::{BufRead, BufReader, Read},
    path::Path,
    time::{Duration, UNIX_EPOCH},
};
const ROW: usize = 8 * 1024 * 1024;
const LIST: usize = 300;
fn db(p: &Path) -> Result<Connection> {
    let c = Connection::open_with_flags(p, OpenFlags::SQLITE_OPEN_READ_ONLY)?;
    c.busy_timeout(Duration::from_secs(2))?;
    c.execute_batch("PRAGMA query_only=ON")?;
    Ok(c)
}
fn lines(p: &Path, mut visit: impl FnMut(Value) -> Result<bool>) -> Result<()> {
    let mut f = BufReader::new(File::open(p)?);
    loop {
        let mut b = Vec::new();
        let n = f
            .by_ref()
            .take((ROW + 1) as u64)
            .read_until(b'\n', &mut b)?;
        if n == 0 {
            break;
        }
        if n > ROW {
            return Err(error("Native history item exceeds 8 MiB"));
        }
        match serde_json::from_slice::<Value>(&b) {
            Ok(v) => {
                if v.is_object() && !visit(v)? {
                    break;
                }
            }
            Err(_) if b.last() != Some(&b'\n') => break,
            Err(_) => return Err(error("Unrecognized native JSONL record")),
        }
    }
    Ok(())
}
fn parts(v: &Value) -> String {
    if let Some(s) = v.as_str() {
        return s.to_owned();
    }
    v.as_array()
        .map(|a| {
            a.iter()
                .filter(|p| {
                    ["text", "input_text", "output_text"]
                        .contains(&p["type"].as_str().unwrap_or(""))
                })
                .filter_map(|p| p["text"].as_str())
                .collect::<Vec<_>>()
                .join("\n")
        })
        .unwrap_or_default()
}
fn text100(s: &str) -> String {
    s.chars().take(100).collect()
}
fn messages(entry: &Value, mut emit: impl FnMut(&str, String) -> Result<()>) -> Result<()> {
    let codex = entry["provider"] == "Codex";
    if codex && entry["mode"] == "paginated" {
        let c = db(Path::new(string(entry, "history_db")?))?;
        let mut st=c.prepare("SELECT length(item_json),substr(item_json,1,?),item_type FROM thread_items WHERE thread_id=? AND item_type IN (?,?) ORDER BY rollout_ordinal")?;
        let mut rows = st.query(params![
            (ROW + 1) as i64,
            string(entry, "id")?,
            "userMessage",
            "agentMessage"
        ])?;
        while let Some(row) = rows.next()? {
            let n: i64 = row.get(0)?;
            let raw: String = row.get(1)?;
            let kind: String = row.get(2)?;
            if n > ROW as i64 || raw.len() > ROW {
                return Err(error("Native history item exceeds 8 MiB"));
            }
            let v: Value = serde_json::from_str(&raw)?;
            let agent = kind == "agentMessage";
            let text = if agent {
                v["text"].as_str().unwrap_or("").to_owned()
            } else {
                parts(&v["content"])
            };
            if !text.is_empty() {
                emit(if agent { "assistant" } else { "user" }, text)?;
            }
        }
        return Ok(());
    }
    lines(Path::new(string(entry, "path")?), |v| {
        let message = if codex {
            if v["type"] != "response_item" || v["payload"]["type"] != "message" {
                return Ok(true);
            }
            &v["payload"]
        } else {
            if !v["sessionId"].is_null() && v["sessionId"] != entry["id"]
                || v["isSidechain"] == true
                || v["isMeta"] == true
                || !["user", "assistant"].contains(&v["type"].as_str().unwrap_or(""))
            {
                return Ok(true);
            }
            &v["message"]
        };
        let role = message["role"].as_str().unwrap_or("");
        if ["user", "assistant"].contains(&role) {
            let text = parts(&message["content"]);
            if !text.is_empty() {
                emit(role, text)?
            }
        }
        Ok(true)
    })
}
fn list(q: &Value) -> Result<Value> {
    let root = Path::new(string(q, "root")?);
    let provider = string(q, "provider")?;
    let dir = q["directory"].as_str();
    let mut entries = Vec::new();
    if provider == "Codex" {
        let c = db(&root.join("state_5.sqlite"))?;
        let mut st=c.prepare("SELECT id,cwd,title,rollout_path,history_mode,updated_at FROM threads WHERE (? IS NULL OR rtrim(cwd,'/')=rtrim(?,'/')) ORDER BY updated_at DESC LIMIT ?")?;
        let mut rows = st.query(params![dir, dir, (LIST + 1) as i64])?;
        while let Some(row) = rows.next()? {
            let mode: String = row.get(4)?;
            if !["legacy", "paginated"].contains(&mode.as_str()) {
                return Err(error("Unsupported Codex history mode"));
            }
            entries.push(json!({"provider":provider,"id":row.get::<_,String>(0)?,"directory":row.get::<_,String>(1)?,"title":text100(&row.get::<_,Option<String>>(2)?.unwrap_or_default()),"path":row.get::<_,Option<String>>(3)?,"mode":mode,"updated":row.get::<_,i64>(5)?,"history_db":root.join("thread_history_1.sqlite")}));
        }
    } else if provider == "Claude" {
        let mut count = 0;
        for project in fs::read_dir(root.join("projects"))? {
            let project = project?.path();
            if !project.is_dir() {
                continue;
            }
            for entry in fs::read_dir(project)? {
                let path = entry?.path();
                if path.extension().is_none_or(|x| x != "jsonl") {
                    continue;
                }
                count += 1;
                if count > 10000 {
                    return Err(error("More than 10000 Claude histories"));
                }
                let (mut cwd, mut title) = (String::new(), String::new());
                let mut index = 0;
                lines(&path, |v| {
                    if cwd.is_empty() {
                        cwd = v["cwd"].as_str().unwrap_or("").to_owned()
                    }
                    if v["type"] == "user" && v["isMeta"] != true {
                        title = text100(&parts(&v["message"]["content"]))
                    }
                    index += 1;
                    Ok((cwd.is_empty() || title.is_empty()) && index <= 100)
                })?;
                if let Some(dir) = dir {
                    if cwd.is_empty() || fs::canonicalize(&cwd).ok() != fs::canonicalize(dir).ok() {
                        continue;
                    }
                }
                let id = path.file_stem().unwrap().to_string_lossy().into_owned();
                let updated = path
                    .metadata()?
                    .modified()?
                    .duration_since(UNIX_EPOCH)?
                    .as_secs_f64();
                entries.push(json!({"provider":provider,"id":id,"directory":cwd,"title":if title.is_empty(){id}else{title},"path":path,"updated":updated}));
                entries.sort_by(|a, b| {
                    b["updated"]
                        .as_f64()
                        .partial_cmp(&a["updated"].as_f64())
                        .unwrap_or(std::cmp::Ordering::Equal)
                });
                entries.truncate(LIST + 1);
            }
        }
    } else {
        return Err(error("Unsupported provider"));
    }
    let truncated = entries.len() > LIST;
    entries.truncate(LIST);
    Ok(json!({"entries":entries,"truncated":truncated}))
}
pub fn dispatch(q: &Value) -> Result<Value> {
    if q["action"] == "current" {
        return Ok(json!({"entry":crate::usage::history_entry(Path::new(string(q, "session")?))?}));
    }
    if q["action"] == "list" {
        return list(q);
    }
    let entry = &q["entry"];
    let limit = q["limit"].as_u64().unwrap_or(65536).clamp(128, 65536) as usize;
    let offset = q["offset"].as_u64().map(|x| x as usize);
    let prompt = q["prompt"] == true;
    let mut total = 0;
    let mut shown = String::new();
    messages(entry, |role, text| {
        if prompt {
            if role == "user" {
                total = text.chars().count();
                shown = text
                    .chars()
                    .skip(offset.unwrap_or(0).min(total))
                    .take(limit)
                    .collect();
            }
            return Ok(());
        }
        let block = format!(
            "\n## {}\n\n{}\n",
            if role == "user" {
                "사용자"
            } else {
                "응답"
            },
            text
        );
        let n = block.chars().count();
        if let Some(o) = offset {
            let a = o.saturating_sub(total);
            let b = n.min(o.saturating_add(limit).saturating_sub(total));
            if b > a {
                shown.extend(block.chars().skip(a).take(b - a));
            }
        } else {
            let tail: String = block.chars().skip(n.saturating_sub(limit)).collect();
            shown.push_str(&tail);
            let n = shown.chars().count();
            shown = shown.chars().skip(n.saturating_sub(limit)).collect();
        }
        total += n;
        Ok(())
    })?;
    let count = shown.chars().count();
    let start = if prompt {
        offset.unwrap_or(0).min(total)
    } else {
        offset.map_or(total.saturating_sub(count), |x| x.min(total))
    };
    Ok(
        json!({"text":shown,"start":start,"end":start+count,"total":total,"provider":entry["provider"],"id":entry["id"],"prompt":prompt}),
    )
}
